### DISCUSSION TITLE
Extension cannot launch fortls

### DISCUSSION THREAD

**Discussion by BambOoxX** [krvajal/vscode-fortran-support#502] [Q&A - deprecated see Fortran Discourse] [answered]
**Describe the bug**
The extension show a pop-up message with 
> Error launching fortls: Please check that all selected options are supported by your language server version.

However, no specific options (except for `fortls` path) have been set (which may be the problem) but I can't see a list of them. 
Searching the output of Modern Fortran logs shows
>["INFO" - 14:24:15] Extension Name: Modern Fortran
["INFO" - 14:24:15] Extension Version: 3.0.202204291734
["INFO" - 14:24:15] Linter set to: ifort
["INFO" - 14:24:15] Formatter set to: findent
["INFO" - 14:24:15] Autocomplete set to: fortls
["INFO" - 14:24:15] Hover set to: fortls
["INFO" - 14:24:15] Symbols set to: fortls
["INFO" - 14:24:15] using linter: ifort located in: ifort
["INFO" - 14:24:15] Linter.arguments:
-warn
all
["INFO" - 14:24:15] Linter.include:

>["INFO" - 14:24:15] Fortran Language Server
["INFO" - 14:25:01] using linter: ifort located in: ifort
["INFO" - 14:25:01] Linter.arguments:
-warn
all

The problem (or at least its symptoms) seems to be close to https://stackoverflow.com/q/72153608/9576551, however I can't see how https://gnikit.github.io/fortls/editor_integration.html#visual-studio-code can help understand which options to change to make `fortls` work.

**Build info (please complete the following information):**

- OS: Windows 10 20H2
- Extension Version v3.0.2022042917
- Visual Studio Code Version 1.67.2
- `fortls` version 2.7.0 installed with `conda`

**Additional context**

**Comment by gnikit** [comment]
Do you have FORTRAN Intellisense installed?

**Comment by BambOoxX** [comment]
No, I do not, given the anwer in https://stackoverflow.com/a/72336810/9576551 I figured it was not required.

**Comment by gnikit** [comment]
It is not, I am just trying to figure out what options are being passed. Could you post your `fortran.` settings from settings.json? Something seems to be out of sync.
I will convert this to a Q&A since this is not a bug but related to your configuration.

**Comment by BambOoxX** [comment]
The only options in my `settings.json` matching `fortran.` are
```
"fortran.fortls.path": "C:\\Anaconda3\\pkgs\\fortls-2.7.0-py38haa244fe_0\\Scripts\\fortls",
"fortran.linter.compiler": "ifort",
```

**Reply by gnikit** [reply]
Shouldn't the end of your fortls.path be `...\\fortls.exe`?

**Reply by BambOoxX** [reply]
Actually, it does not make a difference, when started from an anaconda prompt. However, if I use this path in a conventional windows terminal, it fails with 
>Unable to create process using 'C:\Anaconda3\pkgs\fortls-2.7.0-py38haa244fe_0\python.exe C:\Anaconda3\pkgs\fortls-2.7.0-py38haa244fe_0\Scripts\fortls-script.py '

That behavior may explain why it does not work in the extension...
Why is the exe prepending a path before the python ?

**Reply by gnikit** [reply]
So what happens when you are in a terminal and do something like

```sh
C:\\Anaconda3\\pkgs\\fortls-2.7.0-py38haa244fe_0\\Scripts\\fortls.exe --debug_filepath your-fortran-file.f90 --debug_rootpath . --hover_signature --use_signature_help --debug_parser
```

**Reply by BambOoxX** [reply]
The same error actually, my test was done with `C:\\Anaconda3\\pkgs\\fortls-2.7.0-py38haa244fe_0\\Scripts\\fortls.exe --help`.
If I do a `python C:\Anaconda3\pkgs\fortls-2.7.0-py38haa244fe_0\Scripts\fortls-script.py --help` inside an anaconda prompt, it gets me the help correctly. 
It seems very much related to the `python` call. Is there a way to pass the actual python path or instruction ? 

In order to start an anaconda prompt, I actually need to do `%windir%\System32\cmd.exe "/K" C:\Anaconda3\Scripts\activate.bat C:\Anaconda3` and then I can lauch a `python` inside the anaconda environment

**Reply by gnikit** [reply]
From the looks of it, something is wrong with your anaconda installation. You definetly should not be calling the `fortls-script.py` explicitly, the fortls.exe is made by conda-forge for exactly this reason. I don't think there is a safe way of doing this. You might be able to hack things by changing the fortls path but I am not sure. Either way, sorting this out seems like a nightmare. It could be your fortls installation through conda, it could be your python env, it could also be the underlying anaconda env not being properly initialised, or any combination of the above.


I would recommend just installing `fortls` with `pip` and just forget about it. It will fetch its own updates and you will not have to worry again.

**Comment by gnikit** [comment, marked as answer]
where you able to solve this?

**Reply by BambOoxX** [reply]
Sorry, I had to handle some urgent stuff before coming back to this. I can actually launch the anaconda shell with `%windir%\System32\cmd.exe "/K" C:\Anaconda3\Scripts\activate.bat C:\Anaconda3` then launch `fortls` directly, so no need to launch python then the .py file. 

Th problem is that I can't provide the sequence prior `fortls` instead of what the extension provides.

**Reply by gnikit** [reply]
As I previously mentioned, you are not meant to activate an Anaconda env like that, see: 
- https://docs.conda.io/projects/conda/en/latest/user-guide/getting-started.html#starting-conda
- [win docs extract 1](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#:~:text=On%20Windows%2C%20PATH,multi%2Duser%20installs.) and a bit further down [win docs extract 2](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#:~:text=Windows%20is%20extremely%20sensitive%20to%20proper%20activation.%20This%20is%20because%20the%20Windows%20library%20loader%20does%20not%20support%20the%20concept%20of%20libraries%20and%20executables%20that%20know%20where%20to%20search%20for%20their%20dependencies%20(RPATH).%20Instead%2C%20Windows%20relies%20on%20a%20dynamic%2Dlink%20library%20search%20order.)

I am afraid that your problem is  very specific to your Anaconda installation and I doubt that I will be able to provide any more insights other than "uninstall and reinstall your Anaconda and add it to your PATH", so instead of that I would strongly recommend you
use `pip` **outside of Anaconda** to install `fortls`, that is the most robust way to launch the server.

As a side note, executing a sequence of commands like you mentioned, is extremely unsafe and a major security risk so you should avoid it all costs.

**Reply by BambOoxX** [reply]
>As I previously mentioned, you are not meant to activate an Anaconda env like that, see : [...]

Starting the anaconda prompt actually calls the line `%windir%\System32\cmd.exe "/K" C:\Anaconda3\Scripts\activate.bat C:\Anaconda3` I just copy-pasted the path to make clear what Anaconda prompt really does. The `activate.bat` file is created by Anaconda upon install. 

I agree this seems very much Anaconda related. However, I cannot use `pip` outside of Anaconda, and it's a bad idea to mix `pip` and `conda` for handling packages. Overall, I also have little flexibility due to my company's IT policy.

Do you know of previous examples of `fortls` using Anaconda ?

**Reply by gnikit** [reply]
>Do you know of previous examples of fortls using Anaconda ?

Yes, there is a large number of users that rely on Anaconda and AFAIK they have not had any problems since the first couple `fortls` conda releases. Personally, whenever I use conda I am able to call `fortls` directly e.g. `anaconda3/envs/myenv/bin/fortls` without initialising the python env. I suspect that works because on Linux a bunch of the Python shared libraries are installed globally.

1. Are you able to add things to your PATH (User variables not global)?
2. Are you able to install things on a user level (no admin access)? The answer to this should almost definitely be yes.

Let me know about the 2 points and I will think about a solution

**Reply by BambOoxX** [reply]
> Yes, there is a large number of users that rely on Anaconda and AFAIK they have not had any problems since the first couple fortls conda releases.

Nice, that means it's solvable ^^. 

>Personally, whenever I use conda I am able to call fortls directly e.g. anaconda3/envs/myenv/bin/fortls without initialising the python env. I suspect that works because on Linux a bunch of the Python shared libraries are installed globally.

That is probably the case indeed, python is not always playing nice on windows ... In my case, if I call `fortls` directly, I get the error mentioned above. 
The only way for me to get it to work at the moment is to call e.g. `conda activate base & fortls --help`, so it is really about having the conda environment set up. 

I can add stuff to my path, but it doesn't seem to change much the behavior, since the problem does not come from the lack of access to some exe, rather than to the activation of the environment.
I'm also able to install stuff on my user domain.

**Reply by gnikit** [reply]
I just verified this on a Windows 10 machine with nothing custom in its PATH other than a miniconda3 installation and it works as expected, even within virtual environments that are not initialised, 

```json
  "fortran.fortls.path": "C:\\Users\\gn\\miniconda3\\envs\\fortls\\Scripts\\fortls.exe",
```

You shouldn't have to initialise or activate conda for `fortls` to work, it's just a binary, there is nothing more to it. The reason why it wasn't working for you is because you were calling the binary from `pkgs` which is not what you are supposed to do if you want your binary to work outside of a Python env

**Reply by BambOoxX** [reply]
Yes, that seems to do it indeed, thanks very much for the help ! Have a great day !
